Dm 168 add unsubscription link to the sample update emails in sample notificator cli

KochTobi opened the pull request:
This PR moves the logic of sending emails for project updates into the business domain and adds tests.
The emails sent contain an unsubscription link now.

KochTobi committed:
Rename ds to dataSource

KochTobi committed:
Empty notification lists are not longer forwarded but consumed by the use case

KochTobi committed:
remove duplicate test

KochTobi committed:
Put email generation logic into business domain

KochTobi committed:
Remove unnecessary test

The removed test was covered by the parameterized version

KochTobi committed:
Cleanup invalid html template. Add unsubscription link.

KochTobi committed:
Adjust email html title

KochTobi committed:
Use valid title tag for html template

KochTobi committed:
Adjust email template

Co-authored-by: Jennifer Bödker <jennifer.boedker@qbic.uni-tuebingen.de>

KochTobi committed:
Implement interfaces

KochTobi committed:
Clean up DependencyManager

KochTobi committed:
Instantiate refactored classes

KochTobi committed:
Upgrade Dependencies

KochTobi committed:
Implement Email generation

still missing working token generation

KochTobi committed:
Fix unsubscription request

KochTobi committed:
modify ignored integration test

KochTobi committed:
expand the tests to account for token generation endpoint

KochTobi committed:
wrangle link generation in there

KochTobi committed:
Add missing repository

KochTobi committed:
Remove debug behaviour

KochTobi committed:
Ignore SenderSpecs

jenniferboedker commented on sample-notificator-app/src/main/resources/notification-template/email-update-template.html:
```diff
@@ -1,31 +1,39 @@
+<html lang="en">
+<head>
+  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
+  <title>Sample Status Update</title>
```
Nice, I like the overhauled email template

jenniferboedker commented on sample-notificator-domain/src/main/groovy/life/qbic/business/notification/refactor/FailureEmail.java:
```diff
@@ -0,0 +1,7 @@
+/**
+ * <b>An email informing about system failure.</b>
+ */
+public interface FailureEmail {
```
This is an empty class?

KochTobi replied:
Yes this interface is used to achieve type boxing.

KochTobi replied:
You were right, it turns out I introduced it and forgot to remove it as it was not used in the end.

jenniferboedker commented on sample-notificator-domain/src/test/groovy/life/qbic/business/notification/refactor/SendEmailSpec.groovy:
```diff
@@ -0,0 +1,69 @@
+
+  EmailGenerator emailGenerator = Stub()
+  EmailSender flawlessEmailSender = Mock()
+  FailureEmailSender adminInformer = Mock()
```
Why cant the admin informer not also be an emailsender? but he could just have another type of emails, no NotificationContent but something else?

KochTobi replied:
The adminInformer is an email sender `FailureEmailSender`. As we need the method that this interface provides it needed to be more specific.

KochTobi replied:
Originally, this was the idea. I introduced a NotificationEmail and a FailureEmail. As it turned out the process of constructing and sending those emails was quite different and they did not have anything in common except being called email. Thus I simplified the code which made the FailureEmail not being used any longer and allowed for different interfaces with more explicit methods.
I would like to keep this as is.

jenniferboedker replied:
Sure, I just wanted to understand

jenniferboedker commented on sample-notificator-domain/src/test/groovy/life/qbic/business/notification/refactor/SendEmailSpec.groovy:
```diff
@@ -0,0 +1,69 @@
+  NotificationEmail emailMock = Mock()
+
+  def "when #n notifications are present then send #n emails"() {
+    given: "a notification Mcontent"
```
typo here?

jenniferboedker replied:
```suggestion
    given: "a notification content"
```

jenniferboedker reviewed (changes requested):
Thanks for taking time to refactor the code and give the project more structure. However, I was a little confused about the naming at some places and the function of some elements, maybe we can discuss this in a quick meeting?

KochTobi commented on sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/DependencyManager.groovy:
```diff
@@ -50,34 +55,43 @@ class DependencyManager {
+
+        UnsubscriptionLinkSupplier unsubscriptionLinkSupplier = setupUnsubscriptionLinkSupplier()
+
+        EmailSender<NotificationEmail> emailSender = new HtmlEmailSender()
+        FailureEmailSender failureEmailSender = new SupportEmailSender()
```
The support email is no Html email. It only contains a predefined text

KochTobi committed:
Correct typo

Co-authored-by: jnnfr <jennifer.boedker@qbic.uni-tuebingen.de>

KochTobi commented on sample-notificator-app/src/test/groovy/life/qbic/samplenotificator/components/email/html/HtmlNotificationEmailSpec.groovy:
```diff
@@ -0,0 +1,53 @@
+import life.qbic.business.notification.unsubscription.UnsubscriptionLinkSupplier
+import spock.lang.Specification
+
+/**
+ * <b>short description</b>
+ *
+ * <p>detailed description</p>
+ *
+ * @since <version tag>
+ */
```
Well those are specs so the description would be `tests HtmlNotificationEmail functionality` which is pretty obvious I think.

jenniferboedker reviewed (commented):
small comment before I forget: is it intended that one of the folders is called "refactor"?

jenniferboedker commented on sample-notificator-domain/src/main/groovy/life/qbic/business/notification/refactor/SendEmail.java:
```diff
@@ -0,0 +1,51 @@
+ * <p>Implements the business logic of sending emails for project updates.</p>
+ */
+public class SendEmail implements Consumer<List<NotificationContent>>, SendEmailInput{
+    private final EmailSender<NotificationEmail> sendOrRemember;
+    private final FailureEmailSender informAdmin;
+    private final EmailGenerator<NotificationEmail> generateEmail;
```
Since we have a lot of different names for the same thing I would suggest to not introducing new names here.
The interface is for the use case, right? I find it confusing if the name of the interface is different and implies functionality that is not implied by the interface name.

KochTobi replied:
I would like to keep the class names as is as I think it is reasonably understandable. In case you were referring to the variable names; I chose them so that the code reads like a sentence and they contain even more specific meaning than their class names. The `informAdmin` for example only hints towards what the role of the failure email sender is in this process/pipeline. It still sends failure emails. The same is true for the `sendOrRemember` as it highlights that this sends emails or remembers them for later. The `EmailSender` still sends the email regardless of whether it's remembered emails are of interest. The `generateEmail` is pretty self-explanatory and an `EmailGenerator` should generate emails IMHO.

jenniferboedker replied:
I see why you did it but from my point of view I dont think that it is such a good idea because it seems to me that by naming it like this you make assumptions about what the implementing class does. And if the concret implementation actually is what the class does I dont see why its not reasonable to name the class like this. But anyhow, I will not block this PR because of that.

jenniferboedker commented on sample-notificator-domain/src/main/groovy/life/qbic/business/notification/refactor/NotificationEmail.java:
```diff
@@ -0,0 +1,12 @@
+ * <p>An email containing information of a NotificationContent</p>
+ */
+public interface NotificationEmail {
+  void fill(NotificationContent content);
+  String body();
+  String recipient();
```
Since this is an interface what do you think of adding java doc here?

jenniferboedker commented on sample-notificator-domain/src/main/groovy/life/qbic/business/notification/refactor/SendEmail.java:
```diff
@@ -0,0 +1,51 @@
+ * <p>Implements the business logic of sending emails for project updates.</p>
+ */
+public class SendEmail implements Consumer<List<NotificationContent>>, SendEmailInput{
+    private final EmailSender<NotificationEmail> sendOrRemember;
```
What are reasons for emails not being sent? Would an unstable internet connection or so be reasons e.g. is it software related or are there other reasons?

KochTobi committed:
Rename package refactor to send

Rename `life.qbic.business.notification.refactor` to `life.qbic.business.notification.send`

KochTobi committed:
Rename package refactor to send

Missed some occurences.
Rename `life.qbic.business.notification.refactor` to `life.qbic.business.notification.send`

KochTobi committed:
Remove unused FailureEmail.java

KochTobi committed:
Add JavaDoc to interface

KochTobi committed:
Improve code quality as suggested by analysis

jenniferboedker commented on sample-notificator-domain/src/main/groovy/life/qbic/business/notification/send/SendEmail.java:
```diff
@@ -0,0 +1,51 @@
+ */
+public class SendEmail implements Consumer<List<NotificationContent>>, SendEmailInput{
+    private final EmailSender<NotificationEmail> sendOrRemember;
+    private final FailureEmailSender informAdmin;
```
One small comment again on the naming. What do you think of renaming like this:

```suggestion
    private final AdminEmailSender informAdmin;
```
```suggestion
    private final FailureEmailSender failureEmail;
```
And the same then also in the DependencyManager. That would reduce the complexity and makes it less confusing.

KochTobi committed:
Add JavaDoc

KochTobi committed:
Rename variable to make it application level independent

jenniferboedker commented on sample-notificator-app/src/main/resources/config.properties.example:
```diff
@@ -2,4 +2,9 @@ mysql.host = ...
\ No newline at end of file
+mysql.port = ...
+services.subscriptions.password = ...
+services.subscriptions.tokengeneration.endpoint = ...
```
```suggestion
services.tokengeneration.endpoint = ...
```

jenniferboedker replied:
Needs to be the same as in dependency manager

KochTobi replied:
good catch. I adjusted it in the DependencyManager

jenniferboedker reviewed (changes requested):
When I tested mit @Steffengreiner we found that the url is not correctly set up. There is one '/' too much:
http://localhost:8090/cancel/?token=lajsdflöajsdf
should be
http://localhost:8090/cancel?token=lajsdflöajsdf

KochTobi commented:
> 

You need to set the endpoint in the properties without a tailing `/`. Then it works 😉

KochTobi committed:
Use correct property

Steffengreiner commented:
I think it would also make a lot of sense to update the readme with the now necessary changed conditions: 
1.) Setup subscription service locally 
2.) Adapt `config.properties` to localhost
3.) Maybe add an extra section for adaptions necessary for deployment?

KochTobi committed:
Update README.md

KochTobi committed:
Update README.md

KochTobi commented:
I added branch-specific documentation and confirmed that it works correctly.

KochTobi commented:
Thank you @Steffengreiner and @jenniferboedker for detailed feedback and comprehensive testing :+1:
